tests/js/spec/redux/reducers/shared.spec.js

```diff
@@ -0,0 +1,234 @@
+      listViewState: {
+        ...successState.listViewState,
+        allVisibleSelected: true,
+        selectedIds: new Set(['1', '2']),
+      },
+    };
+    expect(selectedPageState).toEqual(expectedSelectedPageState);
+    const unSelectedPageState = reducer(selectedPageState, actions.selectPage(false));
+
+    expect(unSelectedPageState).toEqual({
+      ...expectedSelectedPageState,
+      listViewState: {
+        ...expectedSelectedPageState.listViewState,
+        allVisibleSelected: false,
+        selectedIds: new Set(),
+      },
+    });
+  });
+
+  it('has expected state when selecting/deselecting a single entry', () => {
+    const successState = createGetListSuccessState();
+    const selectedEntryState = reducer(successState, actions.select('1', true));
+    const expectedSelectedEntryState = {
+      ...successState,
+      listViewState: {
+        ...successState.listViewState,
+        selectedIds: new Set(['1']),
+      },
+    };
+    expect(selectedEntryState).toEqual(expectedSelectedEntryState);
+
+    const unSelectedEntryState = reducer(selectedEntryState, actions.select('1', false));
+    expect(unSelectedEntryState).toEqual({
+      ...selectedEntryState,
+      listViewState: {
+        ...successState.listViewState,
+        selectedIds: new Set(),
+      },
+    });
+  });
+
+  it('has expected state when requesting an entry to be created', () => {
+    const successState = createGetListSuccessState();
+    const createEntryRequestState = reducer(
+      successState,
+      actions.createRequest({id: '1'})
+    );
+    expect(createEntryRequestState).toEqual({
+      ...successState,
+      creating: true,
+    });
+  });
+
+  it('has expected state when requesting an entry has successfully been created', () => {
+    const newItem = {
+      id: '3',
+      name: 'entry3',
+    };
+    const successState = createGetListSuccessState();
+    const createEntryRequestState = reducer(successState, actions.createRequest(newItem));
+    const createEntrySuccessState = reducer(
+      createEntryRequestState,
+      actions.createSuccess(newItem)
+    );
+
+    expect(createEntrySuccessState).toEqual({
+      ...successState,
+      byIds: {
+        ...successState.byIds,
+        '3': newItem,
+      },
+    });
+  });
+
+  it('has expected state after requesting a single entry', () => {
+    const requestedState = reducer(initialState, actions.getRequest());
+    expect(requestedState).toEqual({
+      ...initialState,
+      loadingDetails: true,
+    });
+  });
+
+  it('has expected state when getting a successful single entry response', () => {
+    const requestedState = reducer(initialState, actions.getRequest());
+    // We must get an item with an id back:
+    const fetchedItem = {
+      id: 1,
+    };
+    const successState = reducer(requestedState, actions.getSuccess(fetchedItem));
+    expect(successState).toEqual({
+      ...initialState,
+      detailsId: 1,
+      byIds: {1: {id: 1}},
```

Comment:
Shouldn't visibleIds be updated as well?

Reply:
Good question, but I don't believe it should. When one executes a search or similar one gets two things:

* A partial list (page) of data that's cached in byIds. Currently, byIds will be cleared, but in the future we might selectively keep it for faster paging back and forth or something like that.
* A refreshed list of visible IDs (because we're viewing it in the list view). (An argument could be made to have that view state entirely in the view components themselves, but at the moment we include them in the redux store. If we wouldn't have it in redux, this would be easier to understand I think)

However, when we fetch a single entry, e.g. by entering the page localhost/lab/substance/1, we want to add it to the byIds, so it doesn't need to be looked up again (at least for a while), but we don't want it to pop up in the visible list, because that all depends on if it should be in the current page of data. So if the list has populated samples in alphabetical order [sample1, sample2, sample3] and we suddenly get sample1000, we don't want it to show up in the list. That's up to other actions to decide.

Reply:
Ohh, that was quite an elaborate answer! You are probably right. Fetching a single entity is typically not for showing in a list, rather showing a detailed page for that entity.

Reply:
Yeah, I got carried away there :D